clarity_ext/dilution.py

```diff
@@ -242,11 +250,84 @@ def calculate_transfer_volumes(self):
                     max(transfer.requested_volume - transfer.sample_volume, 0)
                 transfer.has_to_evaporate = \
                     (transfer.requested_volume - transfer.sample_volume) < 0
+                if self.scale_up_low_volumes and transfer.sample_volume < ROBOT_MIN_VOLUME:
+                    scale_factor = float(
+                        ROBOT_MIN_VOLUME / transfer.sample_volume)
+                    transfer.sample_volume *= scale_factor
+                    transfer.buffer_volume *= scale_factor
+                    transfer.scaled_up = True
             except (TypeError, ZeroDivisionError) as e:
                 transfer.sample_volume = None
                 transfer.buffer_volume = None
                 transfer.has_to_evaporate = None
 
+    def split_up_high_volume_rows(self):
+        for transfer in self.transfers:
+            calculation_volume = max(
+                self._get_volume(transfer.sample_volume), self._get_volume(transfer.buffer_volume))
+            (n, residual) = divmod(calculation_volume, PIPETTING_MAX_VOLUME)
+            if residual > 0:
+                duplication_number = int(n + 1)
+            else:
+                duplication_number = int(n)
+
+            # Copy transfer x times and add to array
+            # Update sample volume and buffer volume if needed
+            splitted_transfers = self._split_up_single_transfer(
+                transfer, duplication_number)
+            self.transfers += splitted_transfers
+
+    @staticmethod
+    def _split_up_single_transfer(transfer, number_rows):
+        transfers = []
+        for i in range(0, number_rows):
+            if i == number_rows - 1:
+                t = transfer
+            else:
+                t = copy.copy(transfer)
+                t.buffer_volume = 0
+                t.sample_volume = 0
+                transfers.append(t)
+
+            # Only split up pipetting volume if the actual volume exceeds
+            # max of 50 ul. Otherwise, the min volume of 2 ul might be violated.
+            if transfer.buffer_volume > PIPETTING_MAX_VOLUME:
+                t.buffer_volume = float(
+                    transfer.buffer_volume / number_rows)
+
+            if transfer.sample_volume > PIPETTING_MAX_VOLUME:
+                t.sample_volume = float(
+                    transfer.sample_volume / number_rows)
+
+        return transfers
+
+    def sort_transfers(self):
+        def pipetting_volume(transfer):
+            return self._get_volume(transfer.buffer_volume) + self._get_volume(transfer.sample_volume)
+
+        def max_added_pip_volume():
+            volumes = map(lambda t: (self._get_volume(t.buffer_volume),
+                                     self._get_volume(t.sample_volume)), self.transfers)
+            return max(map(lambda v: v[0] + v[1], volumes))
```

Comment:
For readability, I prefer to use unpacking rather than indexing in statements like these, since it will make it clear to the reader what these are, e.g.:

`lambda buffer_vol, sample_vol: buffer_vol + sample_vol, volumes`